[c10d] default backend need to check for nccl availability

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102336
* __->__ #102470

As titled, we can only initialize nccl backend when NCCL is available

wanchaol committed:
[c10d] default backend need to check for nccl availability

As titled, we can only initialize nccl backend when NCCL is available

[ghstack-poisoned]

Skylion007 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -255,10 +255,9 @@ def __init__(self, backend: Union[str, Backend]):
-                "cpu": Backend.GLOO,
-                "cuda": Backend.NCCL,
-            }
+            self.device_backend_map["cpu"] = Backend.GLOO
```
```suggestion
            self.device_backend_map = {"cpu"] : Backend.GLOO}
```
is marginally faster

wanchaol committed:
Update on "[c10d] default backend need to check for nccl availability"


As titled, we can only initialize nccl backend when NCCL is available

[ghstack-poisoned]

XilunWu reviewed (approved):
LGTM!
